test/app-tap/gh-4450-log-rotate-exit.test.lua

```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env tarantool
+
+local tap = require('tap')
+local test = tap.test("gh-4450-log-rotate-exit")
+local fio = require('fio')
+local tarantool_bin = arg[-1]
+
+test:plan(1)
+
+local function run_script(code)
+    local dir = fio.tempdir()
+    local script_path = fio.pathjoin(dir, 'gh-4450-script.lua')
+    local script = fio.open(script_path,
+        {'O_CREAT', 'O_WRONLY', 'O_APPEND'},
+        tonumber('0777', 8))
+    script:write(code)
+    script:close()
+    local output_file = fio.pathjoin(fio.cwd(), 'gh-4450-out.txt')
+    local cmd = [[/bin/sh -c 'cd "%s" && "%s" ./gh-4450-script.lua 0> %s 2> %s']]
+    local code = os.execute(
+        string.format(cmd, dir, tarantool_bin, output_file, output_file)
+    )
+    fio.rmtree(dir)
+    local out_fd = fio.open(output_file, {'O_RDONLY'})
+    local output = out_fd:read(100000)
+    out_fd:close()
+    return code, output
+end
+
+--
+-- gh-4382: an error in main script should be handled gracefully,
+-- with proper logging.
+--
+local code = "log = require('log')\n"
+code = code .. "box.cfg{log = 'gh-4550.log'}\n"
+code = code .. "box.error.injection.set('ERRINJ_LOG_ROTATE', true)\n"
+code = code .. "log.rotate()\n"
+code = code .. "os.exit()\n"
+local _, output = run_script(code)
+
+test:diag(output)
+test:ok(not string.find(output, "SEGV_MAPERR"),
```

Comment:
Don't you think this test would be simpler to write in luatest?
You wouldn't have to write your own `run_script` function then.

I'm afraid we have to write all new tests in luatest anyways.

Reply:
I'll try, thanks!

Reply:
Not sure how to handle the exit. Wrote the following
```
local server = require('test.luatest_helpers.server')
local t = require('luatest')

local g = t.group()

g.before_all = function()
    g.dflt = server:new({
        alias = 'dflt',
        box_cfg = { log = 'gh-4550.log' },
    })
    g.dflt:start()
end

g.after_all = function()
    g.dflt:drop()
end

g.test_log_rotate_exit = function()
    g.dflt:exec(function()
        box.error.injection.set('ERRINJ_LOG_ROTATE', true)
        require('log').rotate()
        os.exit()
    end)
end
```
Ie without grepping the output but test simply hangs because of `os.exit()` call inside testing function.

Reply:
> Ie without grepping the output but test simply hangs because of os.exit() call inside testing function.

I guess `exec` can't handle functions that don't return?

You could try to wrap the function in a separate fiber

Reply:
> I guess exec can't handle functions that don't return?

Yup, and `eval` as well. I don't get how fiber would help here? You mean to run fiber inside which would call `os.exit` after its start, like deferred work?

Reply:
Yep. Something like
```lua
server:exec(function() require("fiber").new(function()
    box.error.injection.set('ERRINJ_LOG_ROTATE', true)
    require('log').rotate()
    os.exit()
end) end)
```

Reply:
Doesn't hang on my machine. But you forgot to do `g.dflt:exec` of that function.
```lua
local server = require('test.luatest_helpers.server')
local t = require('luatest')

local g = t.group()

g.before_all = function()
    g.dflt = server:new({
        alias = 'dflt',
        box_cfg = { log = 'gh-4550.log' },
    })
    g.dflt:start()
end

g.after_all = function()
    --already stopped.
    g.dflt:cleanup()
end

g.test_log_rotate_exit = function()
    g.dflt:exec(function()
        require("fiber").new(function()
            box.error.injection.set('ERRINJ_LOG_ROTATE', true)
            require('log').rotate()
            os.exit()
        end)
    end)
    require('fiber').sleep(0.1)
    local logfile = require('fio').pathjoin(g.dflt.workdir, 'gh-4550.log')
    t.assert(g.dflt:grep_log('SEGV_MAPERR', nil,
                                      {filename = logfile}) == nil,
                     "No crash after log rotation and exit")
end
```

Something like this should do?

I asked the team and seems we don't add new -tap tests. Only luatest ones

Reply:
Thanks @sergepetrenko ! I tried this, it doesn't trigger an issue :-(

For example if you modify the code this way
```
[cyrill@grain tarantool.git] git diff
diff --git a/src/main.cc b/src/main.cc
index 6af9724de..2fddfc6ee 100644
--- a/src/main.cc
+++ b/src/main.cc
@@ -141,7 +141,7 @@ on_shutdown_f(va_list ap)
        (void) ap;
        trigger_fiber_run(&box_on_shutdown_trigger_list, NULL,
                          on_shutdown_trigger_timeout);
-       say_logger_free();
+       //say_logger_free();
        ev_break(loop(), EVBREAK_ALL);
        return 0;
 }
@@ -584,6 +584,7 @@ tarantool_free(void)
        memtx_tx_manager_free();
        coll_free();
        systemd_free();
+       say_logger_free();
        fiber_free();
        memory_free();
        random_free();
```
Ie bring error back, then the tap test from the patchset should fail since there `SEGV_MAPERR` happens. With lua test I can't trigger the nil dereference for some reason. Maybe due to timings, not sure...

Reply:
I can't reproduce it with luatest either. My thinking was that increasing `ERRINJ_LOG_ROTATE` usleep to something like `1000000` microseconds would help. But it didn't work, unfortunately.

Another idea was that tap test executes the test case in the main fiber, while luatest `exec` executes it in some fiber serving iproto requests. Maybe this was the failure reason. But seems like it wasnt: a plain test case works both in default and in a new fiber:
`tarantool -e "box.cfg{log='asdf.log'} require'log'.rotate() os.exit()"`
`tarantool -e "require'fiber'.new(function() box.cfg{log='asdf.log'} require'log'.rotate() os.exit() end)"`

Reply:
With `ERRINJ_LOG_ROTATE` usleep set to 1000000, the instance's logs look like:
```
2022-07-22 11:00:29.155 [15650] main/103/default.lua I> set 'listen' configuration option to "\/tmp\/t\/001_app-luatest\/dflt.iproto"
2022-07-22 11:00:29.158 [15650] main I> entering the event loop                     
2022-07-22 11:00:29.208 [15650] main/120/iproto.shutdown I> tx_binary: stopped   
2022-07-22 11:00:30.212 [15650] coio I> log file has been reopened   
```
As you can see, the log rotation is delayed well after the `os.exit()` call, but it doesn't help with triggering the issue anyway.

Reply:
@cyrillos any other ideas?

Reply:
nope so far. i would suggest to take tap test instead, since it known to work, and rework it later if needed at all

Reply:
Have you tried not to use `:exec` at all and just use `server.net_box:call(...)`? You don't need to do everything via `:exec()`.

Reply:
Nope, didn't try. If `call` is not blocking then we might try it. Serge/Vlad could you try it please? To be honest I would simply live it as is, the tap test works as expected and I don't see that many reasons to change it.